doc/design/service-discovery.md

```diff
@@ -0,0 +1,316 @@
+
+* Send ID's over IPC channel and read ABA-counter/ServiceDiscoveryChangeCounter from shared memory
+* Data transfer over shared memory
+  * Basically putting Service Registry to shared memory so everyone can access it (read-only)
+
+Pro:
+
+* Fast data transfer
+* Could replace `InterfacePort`s
+
+Contra:
+
+* Bespoken solution more complex than todays approach (thread-safety required, lock-free mechanism needed)
+* In an optimal case IPC channel communication shall only be used for startup
+
+##### Alternative B: Built-in topic based on `InterfacePort`'s
+
+* Built-in topic approach based on `InterfacePort`'s
+  * Sniff and intercept `CaproMessage`
+* Also see
+  * [overview diagram](diagrams/service-discovery-sequence-diagram.puml)
+  * [sequence diagram](diagrams/service-discovery-overview.puml)
+
+Pro:
+
+* Close to DDS mechanism
+* `InterfacePort` already available
+* Synergies with gateway mechanisms
+* Gateways can benefit from event notification mechanism
+  * No need for a polling-loop anymore
+
+Con:
+
+* Overhead, `Listener` will wake up on every `CaproMessage`
+* Bookkeeping at user-side, just delta of service registry is transferred
+* What will RouDi do if he runs out of memory?
+  * Dimensioning according to max values is not optimal (MAX_INTERFACE_CAPRO_FIFO_SIZE)
+  * Presumably lots of memory needed, e.g. during startup phase when lots of apps will do discovery
+  * A safety-certified middleware based on iceoryx would not use the dynamic discovery feature
+
+Remark:
+
+* Need to filter out your own `CaproMessageType::FIND` requests
+* Extend `dispatchCaproMessage()` with condition variable and notification mechanism
+
+##### Alternative C: Custom thread
+
+* Implement a custom thread in `PoshRuntime`, which polls for specific discovery information using request/response
+
+Pro:
+
+* More flexible, e.g. thread can be used for other functionality in future
+* Re-use req/resp building blocks
+
+Con:
+
+* Not needed by all users per default
+* Polling leads to overhead
+
+##### Alternative D: Discovery Class + Listener
+
+Create a new publisher in RouDi which sends a `ServiceRegistryTopic`. This publisher would be used both to signal a
+change in the service registry and to transmit the service discovery registry. The complete old service registry
+(saved locally) would be compared to the new service registry in a new class, extending the public user API.
+
+Pro:
+
+* Simple and consistent user API for both event-based and sychronous requests
+  * Filtering for `findService` could be done inside the new class
+* No need for a change counter as ABA problem solved by POSH mechanisms
+* Not just delta of service registry is transmitted, but complete service registry info
+* Fast data transport
+* Close to DDS mechanism
+* Could replace `InterfacePort`s
+
+Con:
+
+* New publisher in RouDi needed
+* Partially overlapping features with `InterfacePort`s
+  * Could the `InterfacePort`s be replace by alternative D?
+
```

Comment:
After thinking about this design I have some further cons for all the approaches which are using a variation of subscriber & publisher ports (or @mossmaurice eat your own dogfood ;) ). The reason is always that every app would like to do some service discovery.

- The number of subscribers has to be increased by the number of maximum allowed applications. This means instead of 256 subscribers we require 526 subscribers (since we allow 300 apps).
- The number of subscribers per publisher has to be increased to be at least equal to the maximum number of applications (from 256 to 300)
- Because of @elBoberido scenario from above: https://github.com/eclipse-iceoryx/iceoryx/pull/784/files#r638915043 the `IOX_MAX_CHUNKS_ALLOCATED_PER_PUBLISHER_SIMULTANEOUSLY` must be increased from 8 to maximum allowed applications (300) otherwise the publisher may kill the whole roudi when some application is interrupted and there are no more chunks.
- We require consistency checks for our configuration. Lets say someone sets iceoryx up with 300 apps but only 200 subscribers per publisher. Then only 200 apps can do some service discovery. In the end all the mentions from above have to be verified if the setup is feasible.
- We have to increase the maximum number of condition variable by the maximum number of applications. Since the first 300 will be reserved by the service discovery. This can slow down the WaitSet, when it is used with states and not events, if the user uses the default template arguments. 
- We have to ensure the availability of condition variables. This means when we configure 500 and the users acquires too much then new application are unable to acquire new condition variables which are required for the Listener/Waitset during service discovery. One solution would be another condition variable pool only for service discovery.
- We require a separate memory pool in the shared memory which is reserved solely for the service discovery.

In summary you can say those cons lead to a massive resource overhead (a subscriber alone is 16kb large I think) and implementation overhead. Yes, we can reuse code but at a high price since we have to add checks, resource separations etc to guarantee the availability of resources. The additional condition variable pool and mem pool will also be a lot of additional code which has to be tested as well. And this will be a lot of more code then we save.

Here an additional con for alternative D:
- We have some kind of a priority inversion problem. Let's say we have a high priority application and a low priority application. If the low priority is interrupted during its copy then all the high prio applications cannot receive service registry updates until the low priority application has finished its copy work. This has to be avoided.


With this said I would conclude that reusing our ports for service discovery are out of the question since the cons greatly outweigh the pros.

Reply:
> This means instead of 256 subscribers we require 526 subscribers (since we allow 300 apps)

That's a valid point!

> We require consistency checks for our configuration

This will be required anyway, no matter which alternative we choose.

> We have some kind of a priority inversion problem.

That's not the case with our SoFi or MPMC queue. RouDi (highest prio) will always be able to write data.